Utilities/XrdAdaptor/src/XrdFile.cc

```diff
@@ -248,8 +250,57 @@ XrdFile::read (void *into, IOSize n, IOOffset pos)
     addConnection(ex);
     throw ex;
   }
+  if (n == 0) {
+    return 0;
+  }
+
+  // In some cases, the IO layers above us (particularly, if lazy-download is
+  // enabled) will emit very large reads.  We break this up into multiple
+  // reads in order to avoid hitting timeouts.
+  std::future<IOSize> prev_future, cur_future;
+  IOSize bytesRead = 0, prev_future_expected = 0, cur_future_expected = 0;
+  bool readReturnedShort = false;
+
+  // Check the status of a read operation; updates bytesRead and
+  // readReturnedShort.
+  auto check_read = [&](std::future<IOSize> &future, IOSize expected) {
+    if (!future.valid()) {
+      return;
+    }
+    IOSize result = future.get();
+    if (readReturnedShort && (result != 0)) {
+      edm::Exception ex(edm::errors::FileReadError);
+      ex << "XrdFile::read(name='" << m_name << "', n=" << n
+         << ") remote server returned non-zero length read after EOF.";
+      ex.addContext("Calling XrdFile::read()");
+      addConnection(ex);
+      throw ex;
+    } else if (result != expected) {
+        readReturnedShort = true;
+    }
+    bytesRead += result;
+  };
+
+  while (n) {
+    IOSize chunk = std::min(n, static_cast<IOSize>(XRD_CL_MAX_READ_SIZE));
+
+    // Save prior read state; issue new read.
+    prev_future = std::move(cur_future);
+    prev_future_expected = cur_future_expected;
+    cur_future = m_requestmanager->handle(into, chunk, pos);
+    cur_future_expected = chunk;
+
+    // Wait for the prior read; update bytesRead.
+    check_read(prev_future, prev_future_expected);
```

Comment:
So this works the first time through the loop because `prev_future.valid()` is false and `check_read` drops out immediately?